addons/bus/static/src/workers/election_worker.js

```diff
@@ -0,0 +1,120 @@
+import { Deferred } from "@bus/workers/bus_worker_utils";
+
+export class ElectionWorker {
+    MAIN_TAB_TIMEOUT_PERIOD = 3000;
+
+    constructor() {
+        this.masterTab = null;
+        this.candidates = new Set();
+        this.lastHeartbeat = Date.now();
+        this.electionDeferred = new Deferred();
+        this.masterReplyDeferred = null;
+        this.heartbeatCheckInterval = null;
+        this.heartbeatRequestInterval = null;
+    }
+
+    registerCandidate(messagePort) {
+        this.candidates.add(messagePort);
+        if (this.candidates.size === 1) {
+            this.startElection();
+        }
```

Comment:
That's not enough. We could have "frozen" candidates that were unabled to answer previous elections thus, we could have several candidates but no worker. 

Even if we had an election in progress, we have no guarantee that a master tab will be there at the end of the election. So, after discussing it with  @alexkuhn, we think the correct approach would be:

```js
```suggestion
        this.candidates.add(messagePort);
        await this.electionDeferred;
        if (!this.mainTab) {
            startElection();
        }
```

Reply:
And for the part that keep the election forever to run if there is no result, my idea is we don't  clean the `heartbeatCheckInterval` and `heartbeatRequestInterval` when start a election and let's always have them. In this case, it will broadcast HEARTBEAT_REQUEST when masterTab = null.